design/metadata-handling.md

```diff
@@ -0,0 +1,668 @@
+
+##### The networks specifications
+
+The object for the **networks** section list can be one of:
+
+- **ipv4**: an ipv4 static allocation
+- **ipv4Dhcp**: an ipv4 dhcp based allocation
+- **ipv6**: an ipv6 static allocation
+- **ipv6Dhcp**: an ipv6 dhcp based allocation
+- **ipv6Slaac**: an ipv6 slaac based allocation
+
+The **networks/ipv4** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **ipAddress**: the IP address object
+- **netmask**: the netmask, in an integer format
+- **routes**: the list of route objects
+
+The **networks/ipv4/ipAddress** is an address object containing:
+
+- **start**: the start IP address
+- **end**: The end IP address
+- **subnet**: The subnet in a CIDR notation "X.X.X.X/X"
+- **step**: the step between IP addresses
+
+If the **subnet** is specified, then **start** and **end** are not required and
+reverse, if **start** and **end** are specified, then **subnet** is not required
+
+The **networks/ipv4/routes** is a route object containing:
+
+- **network**: the subnet to reach
+- **netmask**: the mask of the subnet as integer
+- **gateway**: the gateway to use
+- **services**: a list of services object as defined later
+
+The **networks/ipv4Dhcp** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+The **networks/ipv6** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **ipAddress**: the IP address object
+- **netmask**: the netmask, in an integer format
+- **routes**: the list of route objects
+
+The **networks/ipv6Dhcp** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+The **networks/ipv6Slaac** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+##### the services specifications
+
+The object for the **services** section list can be one of:
+
+- **dns**: a dns service with the ip address of a dns server
+
+### The Metal3Data object
+
+The output of the controller would be a Metal3Data object,one per node linking to the
+Metal3DataTemplate object and the associated secrets
+
+The Metal3Data object would be:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Data
+metadata:
+  name: nodepool-1-0
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3DataTemplate
+    name: nodepool-1
+spec:
+  index: 0
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-metadata
+    namespace: default
+  metal3Machine:
```

Comment:
I expected to see a metal3Data reference on the machine instead of having the reference point from the data to machine. What benefit do we get from doing it this way?

Reply:
It's for the template rendering to get it quickly, rather than having to parse the ownerReferences.